server/src/main/java/org/apache/druid/indexing/overlord/IndexerMetadataStorageCoordinator.java

```diff
@@ -632,4 +632,21 @@ List<Interval> getUnusedSegmentIntervals(
    */
   boolean markSegmentAsUsed(SegmentId segmentId);
 
+  /**
+   * Validates the given supervisorId and given metadata to ensure
+   * that start/end metadata non-null implies supervisor ID is non-null.
+   */
+  static void validateDataSourceMetadata(
+      @Nullable final String supervisorId,
+      @Nullable final DataSourceMetadata startMetadata,
+      @Nullable final DataSourceMetadata endMetadata
+  )
+  {
+    if ((startMetadata == null && endMetadata != null) || (startMetadata != null && endMetadata == null)) {
+      throw new IllegalArgumentException("start/end metadata pair must be either null or non-null");
```

Comment:
We should throw `InvalidInput.exception` instead as the exception is user-facing and would be sent back in an HTTP response.